### PR TITLE
[NFC][SYCL] Add pragma unroll in accessor.hpp

### PR DESCRIPTION
**Pull request by MrSidims**
It shall improve performance on various targets in case if we
have 3 dimensional calculations within nested loops.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

**Comment by MrSidims** [comment]
@s-kanaev please take a look

File: sycl/include/CL/sycl/accessor.hpp
Comment (s-kanaev) [inline sycl/include/CL/sycl/accessor.hpp:839]: Why is it needed for device code only?

File: sycl/include/CL/sycl/accessor.hpp
Comment (s-kanaev) [inline sycl/include/CL/sycl/accessor.hpp:838]: @erichkeane , could you, please, clarify if it will work out in device compilation mode?

**Review by s-kanaev — COMMENTED** [review]
Other than the questions the patch looks good.

File: sycl/include/CL/sycl/accessor.hpp
Comment (MrSidims) [reply sycl/include/CL/sycl/accessor.hpp:839]: We guarantee, that we compile device code with clang, where the pragma is known. Host compiler can any C++ compiler and not all of them know the pragma.

Comment (erichkeane) [reply sycl/include/CL/sycl/accessor.hpp:839]: Unknown pragmas are supposed to be ignored (with a warning, though 'system' headers don't warn!), so I don't know if the preprocessing is necessary.  

Additionally, limiting us to device only seems like too big hammer here, I would expect if you're limiting this being there,  you would do it more to detect the right compiler.  Specifically, I know GCC and ICC also support this.

File: sycl/include/CL/sycl/accessor.hpp
Comment (AaronBallman) [reply sycl/include/CL/sycl/accessor.hpp:838]: Does this diagnose the unknown pragma (which may break -Werror builds) when SYCL headers are consumed by a compiler other than Clang, like MSVC?

Comment (MrSidims) [reply sycl/include/CL/sycl/accessor.hpp:838]: Yes, it breaks werror build. That's why I've added device check.

File: sycl/include/CL/sycl/accessor.hpp
Comment (MrSidims) [reply sycl/include/CL/sycl/accessor.hpp]: I've added compiler checks. If it won't work, I guess I'll add unrolling via template magic.

**Review by erichkeane — DISMISSED** [review]
I'm more in favor of this approach, I think it at least is less of a 'big hammer'.

**Review by s-kanaev — COMMENTED** [review]
Should there be a test for it?

**Comment by MrSidims** [comment]
Hah, build for cuda is still failing with:
accessor.hpp:841:0: error: ignoring #pragma GCC unroll [-Werror=unknown-pragmas]
 #pragma GCC unroll 3

that's a bit unexpected, I'll check tomorrow why it happens.

>> Should there be a test for it?

@s-kanaev I can barely imagine a proper test for this patch. One option would be to add a test to check_device_code and see unroll metadata appearing on the appropriate branch instruction, but I guess clang itself has lots of tests for that. Also it doesn't include host code compilation on various compilers. Another option would be to, hm, dump assembly? But I can barely imagine, how we can do it (for example) for MSVC compilation in LIT environment. Or there is another option that you have in mind?

**Review by s-kanaev — APPROVED** [review]
LGTM

File: sycl/include/CL/sycl/detail/defines_elementary.hpp
Comment (AaronBallman) [inline sycl/include/CL/sycl/detail/defines_elementary.hpp:110]: The `_Pragma` operator takes a string literal (http://eel.is/c++draft/cpp.pragma.op#1), so I don't think this is quite correct. Above there is a `__SYCL_GCC_PRAGMA` macro, that strikes me as what you want to use here.

Comment (MrSidims) [reply sycl/include/CL/sycl/detail/defines_elementary.hpp]: Yeah, also there should be a stringification to process 'x' properly. The last patch worked in the compiler explorer at least for all the compilers I claim to have unroll pragma.

File: sycl/include/CL/sycl/detail/defines_elementary.hpp
Comment (AaronBallman) [inline sycl/include/CL/sycl/detail/defines_elementary.hpp:107]: This should not be redefined three different times. Also, if this is the approach you want to take, there's some other uses of the stringize operator that could be replaced in this file (rather than using `__SYCL_GCC_PRAGMA` in some places and this approach in others).

Comment (MrSidims) [reply sycl/include/CL/sycl/detail/defines_elementary.hpp]: Thanks! Applied (guess tho __SYCL_GCC_PRAGMA is the only definition to change in the file)

File: sycl/include/CL/sycl/detail/defines_elementary.hpp
Comment (AaronBallman) [inline sycl/include/CL/sycl/detail/defines_elementary.hpp:120]: Any reason to scope so many other macros on whether `STRINGIFY` was previously defined or not?

Also, I would name it `__STRINGIFY` so it doesn't clash with user-defined macros and I would `#undef` it when you're done using it because we don't want people to rely on accessor.hpp giving them this macro as part of its interface.

File: sycl/include/CL/sycl/detail/defines_elementary.hpp
Comment (AaronBallman) [inline sycl/include/CL/sycl/detail/defines_elementary.hpp:106]:
```suggestion
// Define __SYCL_UNROLL to add pragma/attribute unroll to a loop.
```

File: sycl/include/CL/sycl/detail/defines_elementary.hpp
Comment (AaronBallman) [inline sycl/include/CL/sycl/detail/defines_elementary.hpp:111]: I think the order matters here because `__clang__` and `__INTEL_LLVM_COMPILER` are both defined at the same time. So I think we should check for the Intel compilers *first* and then fall back on Clang. WDYT?

File: sycl/include/CL/sycl/detail/defines_elementary.hpp
Comment (MrSidims) [reply sycl/include/CL/sycl/detail/defines_elementary.hpp:120]: Renamed `STRINGIFY` and moved it a bit later. TBH I think that `__STRINGIFY` can be used again at some point of time in defines_elementary.hpp .

Regarding `#undef` - well, it's a valid comment, but surprisingly it doesn't work, see: https://godbolt.org/z/75nrETGzf

File: sycl/include/CL/sycl/detail/defines_elementary.hpp
Comment (MrSidims) [reply sycl/include/CL/sycl/detail/defines_elementary.hpp:106]: Done, thanks!

File: sycl/include/CL/sycl/detail/defines_elementary.hpp
Comment (MrSidims) [reply sycl/include/CL/sycl/detail/defines_elementary.hpp:111]: Indeed, applied

File: sycl/include/CL/sycl/detail/defines_elementary.hpp
Comment (MrSidims) [reply sycl/include/CL/sycl/detail/defines_elementary.hpp]: So yes,

> but surprisingly it doesn't work

Basically defines_elementary.hpp  is a helper file used across SYCL headers, including accessor.hpp

File: sycl/include/CL/sycl/detail/defines_elementary.hpp
Comment (AaronBallman) [inline sycl/include/CL/sycl/detail/defines_elementary.hpp:93]: Why not `#endif // __STRINGIFY` right here? I don't see why the rest needs to be gated (e.g., if a user defines `__STRINGIFY` before including this file, suddenly they don't get `__SYCL_UNROLL` and that seems surprising.)

Comment (MrSidims) [reply sycl/include/CL/sycl/detail/defines_elementary.hpp:94]: Applied, thanks!

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Review by s-kanaev — COMMENTED** [review]
LGTM

**Comment by bader** [comment]
> LGTM

@s-kanaev, please, approve using GitHub UI to unblock the merge.

**Review by s-kanaev — APPROVED** [review]

File: sycl/include/CL/sycl/detail/defines_elementary.hpp
Comment (bader) [inline sycl/include/CL/sycl/detail/defines_elementary.hpp:93]: This duplicates `__SYCL_STRINGIFY_LINE_HELP`.
Please, update `common.hpp` and `config.cpp` and remove duplicate.

BTW, why `__STRINGIFY` and not `__SYCL_STRINGIFY`?

Comment (MrSidims) [reply sycl/include/CL/sycl/detail/defines_elementary.hpp]: Applied, thanks!

**Review by bader — APPROVED** [review]

**Comment by bader** [comment]
@MrSidims, I think, it's more likely that the failure on Linux machine was introduced by @alexbatashev with https://github.com/intel/llvm/commit/739487cbb0994e006568a584dcf549ebb107bc11 rather than this patch.
@alexbatashev, could you take a look, please?

@romanovvlad, FYI.

**Comment by alexbatashev** [comment]
@bader [this](https://github.com/intel/llvm/pull/4369/files) sneaked in before my patch was merged but after it passed CI. Working on a fix.

File: sycl/include/CL/sycl/detail/defines_elementary.hpp
Comment (keryell) [inline sycl/include/CL/sycl/detail/defines_elementary.hpp:120]: While looking at this I wonder why just using a metaprogrammed loop instead is not better than this pragmatic spaghetti plate.

Comment (aelovikov-intel) [reply sycl/include/CL/sycl/detail/defines_elementary.hpp:120]: It also works with one-iteration loops, unlike this pragma-based solution: https://github.com/intel/llvm/pull/6560

**Pull request merged**
